File: src/textify.go
package main

import (
	"fmt"
	"image"
	"image/jpeg"
	"image/png"
	"log"
	"os"
	"strings"
	"time"

	"github.com/inancgumus/screen"
)

func main() {
	wW, wH := screen.Size()
	fmt.Printf("Window Size: %d, %d\n", wW, wH)

	// Get the input file's path
	pwd, err := os.Getwd()
	ReportError(err)
	inputPath := pwd + "\\res\\" + GetInputMedia()
	// Varify that the file is readable (valid), and get its type (image/sequence)
	inputExt := GetFileExt(inputPath)
	inputType := IsValidAndType(inputExt)
	// Set the encoding for the images
	if inputExt == "jpg" {
		image.RegisterFormat("jpeg", "jpg", jpeg.Decode, jpeg.DecodeConfig)
	} else if inputExt == "png" {
		image.RegisterFormat("png", "png", png.Decode, png.DecodeConfig)
	}
	// Do different actions for each type of input
	if inputType == "!valid" {
		fmt.Printf("%s\n", "The provided input type/extension is not supported, sorry!")
	} else if inputType == "image" {
		frame := GetImage(inputPath)
		fmt.Printf("\r%s", frame)
	} else if inputType == "sequence" {
		// Get goal FPS/Frametime
		var fpsGoal int
		fmt.Print("Set the FPS: ")
		_, err := fmt.Scan(&fpsGoal)
		ReportError(err)
		frameTimeGoal := float32(1) / float32(fpsGoal)
		// Read all elements in the dir
		files, err := os.ReadDir(inputPath)
		ReportError(err)
		// Let the frame processing begin!
		var asciiFrames []string
		var totalProcessingTime int64
		for _, f := range files {
			start := time.Now().UnixMilli()
			framePath := inputPath + "\\" + f.Name()
			frame := GetImage(framePath)
			asciiFrames = append(asciiFrames, frame)
			end := time.Now().UnixMilli()
			totalProcessingTime += end - start
			// add progress bar at some point
			fmt.Printf("\rPath: '%s' | Elapsed Time (sec): %d", framePath, totalProcessingTime/1000)
		}
		// Print each frame with a delay to sustain FPS goal
		// Warning: still can be perminantly behind when (frame time > 1/fps)
		// Warning2: this does not account for the amount of time it takes to actually print the "image"
		for _, f := range asciiFrames {
			start := time.Now().UnixMilli()
			fmt.Printf("\r%s", f)
			end := time.Now().UnixMilli()
			// The frame time goal in seconds minus the frame print time in seconds
			// Warning3: im honestly not entirely sure this works yet
			delay := ((frameTimeGoal) * float32(time.Second)) - float32(end-start)/float32(time.Second)
			time.Sleep(time.Duration(delay))
		}

	} else {
		fmt.Printf("%s\n", "IsValidAndType gave erroneous response. Issue Unknown.")
	}

	fmt.Printf("%s", "Closing in 10 seconds...")
	time.Sleep(time.Duration(10 * time.Second))
}

// ==================== APPLICATION SPECIFIC FUNCTIONS ============================================================================================================================

func GetInputMedia() string {
	// Read all elements in the dir "res"
	files, err := os.ReadDir("res")
	ReportError(err)
	// Print every element
	for i, f := range files {
		fmt.Printf("%d %s\n", i, f.Name())
	}
	// Get file selection from integer input
	fmt.Print("Choose file (#): ")
	var chosenFile int
	_, err = fmt.Scanf("%d", &chosenFile)
	ReportError(err)
	if chosenFile >= len(files) {
		log.Fatal("Selected file does not exist. Number provided is likely too large or small.")
	}
	fmt.Printf("Chosen file: #%d '%s'\n", chosenFile, files[chosenFile].Name())
	// Return the name of the selection
	return files[chosenFile].Name()
}

func IsStringInList(str string, list []string) bool {
	for _, elem := range list {
		if elem == str {
			return true
		}
	}
	return false
}

func GetFileExt(fileName string) string {
	fileExtIndex := strings.Index(fileName, ".")
	if fileExtIndex >= 0 {
		return fileName[fileExtIndex:]
	} else {
		return ""
	}
}

func IsValidAndType(fileExt string) string {
	validImageExts := []string{".jpg"}
	if IsStringInList(fileExt, validImageExts) {
		return "image"
	} else if fileExt == "" {
		return "sequence"
	} else {
		return "!valid"
	}
}

// TODO: OPTIMIZE THE SH*T OUT OF THIS
func GetImage(filePath string) string {
	// Get the size (in characters) of the current window
	_, windowHeight := screen.Size()
	charmap := " .-=+*#%@" //[]rune{' ', '\u2591', '\u2592', '\u2593'}
	// Open the image
	file, err := os.Open(filePath)
	ReportError(err)
	defer file.Close()
	// Load the image into readable form
	img, _, err := image.Decode(file)
	ReportError(err)
	// Get the image's size (bounds)
	bounds := img.Bounds()
	originalWidth, originalHeight := bounds.Max.X, bounds.Max.Y
	// Get the width, adjusted to the terminal's size
	// (*2 because 1 char is double high than wide)
	newWidth := int(float32(originalWidth) / float32(originalHeight) * float32(windowHeight) * 2)
	widthRatio := float32(originalWidth) / float32(newWidth) // image pixels per char "pixel"
	heightRatio := float32(originalHeight) / float32(windowHeight)
	// Loop through every pixel and print the associated char with its luminescence
	var frame string
	for y := float32(0.0); y < float32(originalHeight); y += heightRatio {
		for x := float32(0.0); x < float32(originalWidth); x += widthRatio {
			r, g, b, _ := img.At(int(x), int(y)).RGBA()
			lum := CalcLumine(r, g, b)
			lum = Normalize(lum, 65535, len(charmap)-1)
			frame += string(charmap[lum])
		}
		frame += "\n"
	}
	return frame
}

// ==================== EXTRA UTILITY FUNCTIONS ===================================================================================================================================

func ReportError(e error) {
	if e != nil {
		fmt.Printf("Error: %v", e)
	}
}

func CalcLumine(r, g, b uint32) int {
	return int((r + g + b) / 3)
}

func Normalize(value, oldMax, newMax int) int {
	weight := float32(oldMax) / float32(newMax)
	return int(float32(value) / weight)
}

File: cpp/src/textify.cpp
#define STB_IMAGE_IMPLEMENTATION
#define STB_IMAGE_RESIZE2_IMPLEMENTATION

#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <windows.h>
#include <string>
#include <filesystem>
namespace fs = std::filesystem;
#include <chrono>
namespace sc = std::chrono;
#include <vector>
#include <thread>
#include <tuple>


#include "../lib/STB/stb_image.h"
#include "../lib/STB/stb_image_resize2.h"

#define OUTPUT_CHANNELS 3

const std::string charmap = " .-=+*#%@";

/* Get the Terminal window size in characters */
std::tuple<int, int> GetWindowSize();
/* Force the window to maximize (borderless fullscreen) */
void MaximizeWindow();
/* Get which file will be converted into text. Requires user input as an integer. */
std::string GetInputFile();
/* Simple Util to check if string is in vector. This might be a built in function */
bool IsStringInVec(std::string str, std::vector<std::string> vec);
/* Get the file extension from a file's name/path */
std::string GetFileExt(std::string fileName);
/* Check if the file's extension is valid (as in it is supported by this program & stb) */
int IsValidAndType(std::string fileExt);
/* Get the ascii-converted image from a file's path */
std::string GetImage(std::string filePath);
int CalcLumine(int r, int g, int b);
int Normalize(int val, int oldmax, int newmax);

int main()
{
	if (!fs::exists("res"))
		if (!fs::create_directory("res"))
			perror("Error creating directory");

	MaximizeWindow();

	// Get which file to use for the image
	std::string fileName = GetInputFile();
	std::string filePath = fs::current_path().string() + "\\res\\" + fileName;
	printf("Chosen file path: '%s'\n", filePath.c_str());

	// Varify that the file is readable (valid), and get its type (image/sequence)
	std::string fileExt = GetFileExt(filePath);
	int fileType = IsValidAndType(fileExt);
	// Do different actions for each type of input
	if (fileType == -1) {
		printf("%s\n", "The provided input type/extension is not supported, sorry!");
	} 
	else if (fileType == 0) {
		printf("\r%s", GetImage(filePath));
	}
	else if (fileType == 1) {
		// Get goal FPS/Frametime
		int fpsGoal;
		printf("%s", "Set fps goal: ");
		scanf("%d", &fpsGoal);
		float frameTimeGoal = 1 / fpsGoal;
		// Read all elements in the dir
		std::vector<std::string> files = {};
		for (const auto & entry : fs::directory_iterator(filePath))
			files.emplace_back(entry.path().string());
		// Let the frame processing begin!
		std::vector<std::string> asciiFrames;
		int totalProcessingTime;
		for (auto & file : files) {
			std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
			std::string framePath = filePath + "\\" + file;
			asciiFrames.emplace_back(GetImage(framePath));
			std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
			totalProcessingTime += std::chrono::duration_cast<std::chrono::microseconds>(end - begin).count();
			// add progress bar at some point
			printf("\rPath: '%s' | Elapsed Time (sec): %d", framePath, totalProcessingTime / 1000);
		}
		// Print each frame with a delay to sustain FPS goal
		// Warning: still can be perminantly behind when (frame time > 1/fps)
		// Warning2: this does not account for the amount of time it takes to actually print the "image"
		for (auto & frame : asciiFrames) {
			std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
			printf("\r%s", frame.c_str());
			std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
			// The frame time goal in seconds minus the frame print time in seconds
			// Warning3: im honestly not entirely sure this works yet
			float delay = frameTimeGoal - std::chrono::duration_cast<std::chrono::microseconds>(end - begin).count();
			Sleep(delay);
		}

	} else {
		printf("%s\n", "IsValidAndType gave erroneous response. Issue Unknown.");
	}

	printf("%s", "Closing in 10 seconds...");
	Sleep(100000);
}

std::tuple<int, int> GetWindowSize()
{
	HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
	CONSOLE_SCREEN_BUFFER_INFO csbi;
	GetConsoleScreenBufferInfo(hConsole, &csbi);
	return {csbi.srWindow.Right - csbi.srWindow.Left + 1, 
			csbi.srWindow.Bottom - csbi.srWindow.Top + 1};
}

void MaximizeWindow()
{
	HWND consoleWindow = GetConsoleWindow();
	ShowWindow(consoleWindow, SW_MAXIMIZE);
}

std::string GetInputFile()
{
	std::vector<std::string> files = {};

	for (const auto & entry : fs::directory_iterator("res\\")) {
		std::string entryPath = entry.path().string();
		entryPath = entryPath.substr(4); // remove "res\\"
		files.emplace_back(entryPath);
	}

	// display every file in the directory
	unsigned int i = 0;
	for (auto & filepath : files)
		printf("%d %s\n", i++, filepath.c_str());


	// select one of the displayed files
	int selection;
	printf("What file do you choose?: ");
	scanf("%d", &selection);

	std::string chosenFile = files.at(selection);

	return chosenFile;
}

bool IsStringInVec(std::string str, std::vector<std::string> vec)
{
	for (auto & elem : vec)
		if (elem == str)
			return true;
	
	return false;
}

std::string GetFileExt(std::string fileName)
{
	// get the index of the "." in the filename, and then set ext to everything past it
	int fileExtIndex = fileName.find(".");
	if (fileExtIndex == std::string::npos)
		return "";
	std::string fileExt = fileName.substr(fileExtIndex);
	return fileExt;
}

int IsValidAndType(std::string fileExt)
{
	std::vector<std::string> valid = {".jpg"};
	if (IsStringInVec(fileExt, valid))
		return 0;
	else if (fileExt.empty())
		return 1;
	else
		return -1;
}

std::string GetImage(std::string filePath)
{
	auto [winWidth, winHeight] = GetWindowSize();

	// Load the image file
	int imgWidth, imgHeight, channels;
	unsigned char *image = stbi_load(filePath.c_str(), &imgWidth, &imgHeight, &channels, OUTPUT_CHANNELS);
	if (image == nullptr)
		perror("Error loading image.");

	int newWidth = imgWidth / (imgHeight * winWidth * 2);
	float widthRatio = imgWidth / newWidth;
	float heightRatio = imgHeight / winHeight;

	std::string frame;
	for (float y = 0.0f; y < imgHeight; y += heightRatio) {
		for (float x = 0.0f; x < imgWidth; x += widthRatio) {
			int pixelOffset = (int(y) + newWidth * int(x)) * channels;
			int red = image[pixelOffset + 0];
			int green = image[pixelOffset + 1];
			int blue = image[pixelOffset + 2];
			//int a = channels >= 4 ? image[pixelOffset + 3] : 0xff;
			int lum = CalcLumine(red, green, blue);
			lum = Normalize(lum, USHRT_MAX, charmap.size());
			frame += charmap.at(lum);
		}
		frame += "\n";
	}

	stbi_image_free(image);
	return frame;
}

int CalcLumine(int r, int g, int b){
	return int(0.299 * r + 0.587 * g + 0.114 * b);
}

int Normalize(int val, int oldmax, int newmax) {
	return (val) * (newmax) / (oldmax);
} 